packages/firestore/src/local/indexeddb_persistence.ts

```diff
@@ -285,48 +310,64 @@ export class IndexedDbPersistence implements Persistence {
-            // settings is not supported.
-            //
-            // TODO(multitab): Remove this check when `synchronizeTabs` can no
-            // longer be turned off.
-            throw new FirestoreError(
-              Code.FAILED_PRECONDITION,
-              PRIMARY_LEASE_EXCLUSIVE_ERROR_MSG
-            );
-          }
+          if (!this.isLocalClient(currentPrimary)) {
+            if (!currentPrimary.allowTabSynchronization) {
+              // Fail the `canActAsPrimary` check if the current leaseholder has
+              // not opted into multi-tab synchronization. If this happens at
+              // client startup, we reject the Promise returned by
+              // `enablePersistence()` and the user can continue to use Firestore
+              // with in-memory persistence.
+              // If this fails during a lease refresh, we will instead block the
+              // AsyncQueue from executing further operations. Note that this is
+              // acceptable since mixing & matching different `synchronizeTabs`
+              // settings is not supported.
+              //
+              // TODO(multitab): Remove this check when `synchronizeTabs` can no
+              // longer be turned off.
+              throw new FirestoreError(
+                Code.FAILED_PRECONDITION,
+                PRIMARY_LEASE_EXCLUSIVE_ERROR_MSG
+              );
+            }
 
-          return false;
+            return false;
+          }
         }
 
-        // Check if this client is eligible for a primary lease based on its
-        // visibility state and the visibility state of all active clients. A
-        // client can obtain the primary lease if it is either in the foreground
-        // or if this client and all other clients are in the background.
-        if (this.inForeground) {
+        if (this.networkEnabled && this.inForeground) {
           return true;
         }
 
         let canActAsPrimary = true;
         return clientMetadataStore(txn)
-          .iterate((key, value, control) => {
-            if (this.clientId !== value.clientId) {
+          .iterate((key, otherClient, control) => {
+            if (
+              this.clientId !== otherClient.clientId &&
+              this.isWithinMaxAge(otherClient.updateTimeMs)
+            ) {
+              const otherClientHasBetterNetworkState =
+                !this.networkEnabled && otherClient.networkEnabled;
+              const otherClientHasBetterVisibility =
+                !this.inForeground && otherClient.inForeground;
+              const otherClientHasSameOnlineState =
+                this.networkEnabled === otherClient.networkEnabled;
               if (
-                this.isWithinMaxAge(value.updateTimeMs) &&
-                value.inForeground
+                otherClientHasBetterNetworkState ||
+                (otherClientHasBetterVisibility &&
+                  otherClientHasSameOnlineState)
               ) {
```

Comment:
It's a little tempting to introduce a numeric "primary priority" and have a helper like:

```
  /** Returns the "priority" of a client to become primary.  If any client has a priority higher than ours, then we shouldn't try to become primary. */
  primaryPriority(networkEnabled: boolean, inForeground: boolean): number {
    let priority = 0;
    // Network takes precedence over being in the foreground.
    priority = priority + (networkEnabled ? 1: 0);
    priority = (priority << 1) + (inForeground ? 1: 0);
    return priority;
  }
```

And then in here:
```
  const ourPriority = this.primaryPriority(this.networkEnabled, this.inForeground);
  ...
              if (this.primaryPriority(otherClient.networkEnabled, otherClient.inForeground) > ourPriority) {
                canActAsPrimary = false;
                control.done();
              }
```

I think this would be easier to read (and expand later if necessary) than the conditional logic.